contracts/Tokens/Prime/Prime.sol

```diff
@@ -0,0 +1,635 @@
+    }
+
+    /**
+     * @notice Add a market to prime program
+     * @param vToken address of the market vToken
+     * @param supplyMultiplier the multiplier for supply cap. It should be converted to 1e18
+     * @param borrowMultiplier the multiplier for borrow cap. It should be converted to 1e18
+     */
+    function addMarket(address vToken, uint256 supplyMultiplier, uint256 borrowMultiplier) external {
+        _checkAccessAllowed("addMarket(address,uint256,uint256)");
+        if (markets[vToken].exists) revert MarketAlreadyExists();
+
+        markets[vToken].rewardIndex = 0;
+        markets[vToken].supplyMultiplier = supplyMultiplier;
+        markets[vToken].borrowMultiplier = borrowMultiplier;
+        markets[vToken].sumOfMembersScore = 0;
+        markets[vToken].exists = true;
+
+        vTokenForAsset[_getUnderlying(vToken)] = vToken;
+
+        allMarkets.push(vToken);
+        _startScoreUpdateRound();
+    }
+
+    /**
+     * @notice Set limits for total tokens that can be mined
+     * @param irrevocableLimit total number of irrevocable tokens that can be minted
+     * @param revocableLimit total number of revocable tokens that can be minted
+     */
+    function setLimit(uint256 irrevocableLimit, uint256 revocableLimit) external {
+        _checkAccessAllowed("setLimit(uint256,uint256)");
+        if (irrevocableLimit < _totalIrrevocable || revocableLimit < _totalRevocable) revert InvalidLimit();
+
+        _revocableLimit = revocableLimit;
+        _irrevocableLimit = irrevocableLimit;
+    }
+
+    /**
+     * @notice Directly issue prime tokens to users
+     * @param isIrrevocable is the tokens being issued is irrevocable
+     * @param users list of address to issue tokens to
+     */
+    function issue(bool isIrrevocable, address[] calldata users) external {
+        _checkAccessAllowed("issue(bool,address[])");
+
+        if (isIrrevocable) {
+            for (uint i = 0; i < users.length; i++) {
+                _mint(true, users[i]);
+                _initializeMarkets(users[i]);
+            }
+        } else {
+            for (uint i = 0; i < users.length; i++) {
+                _mint(false, users[i]);
+                _initializeMarkets(users[i]);
+                delete stakedAt[users[i]];
+            }
+        }
+    }
+
+    /**
+     * @notice Executed by XVSVault whenever user's XVSVault balance changes
+     * @param user the account address whose balance was updated
+     */
+    function xvsUpdated(address user) external {
+        uint256 totalStaked = _xvsBalanceOfUser(user);
+        bool isAccountEligible = isEligible(totalStaked);
+
+        if (tokens[user].exists && !isAccountEligible) {
+            address[] storage _allMarkets = allMarkets;
+            if (tokens[user].isIrrevocable) {
+                for (uint i = 0; i < _allMarkets.length; i++) {
+                    executeBoost(user, _allMarkets[i]);
+                    updateScore(user, _allMarkets[i]);
+                }
```

Comment:
Refactor. We could extract this loop to an internal function, and reuse it in the last `else` branch. The `_allMarkets` could be defined inside that internal function.